### PR TITLE
SERVER-27310 zlib compression

### PR DESCRIPTION
**Pull request by bjori**
SERVER-15974 Vendorize compress.c & uncompr.c from zlib 1.2.8 too
    SERVER-27310 Add support for zlib wire protocol compression

### CODE REVIEW

**Comment by FrostKiwi** [comment]
Astronomical impact. If reporting is accurate, this caused https://nvd.nist.gov/vuln/detail/CVE-2025-14847 , with which Ubisoft just got compromised, with a successive pivot to an internal git repo and decades of source code exfiltrated.

**Comment by returnofblank** [comment]
LGTM

File: src/mongo/transport/message_compressor_zlib.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                           &length,
+                           reinterpret_cast<const Bytef*>(input.data()),
+                           input.length());
+
+    if (ret != Z_OK) {
+        return Status{ErrorCodes::BadValue, "Compressed message was invalid or corrupted"};
+    }
+
+    counterHitDecompress(input.length(), output.length());
+    return {output.length()};
```
[the remark this answers is not in the document]
Comment (junyeong-huray) [reply src/mongo/transport/message_compressor_zlib.cpp]: The actual fix is just a one-line change.
```diff
-    return {output.length()};
+    return {length};
```
https://github.com/mongodb/mongo/commit/615dc15c17a5c9a0c59ba79bc28f74669e3358cc#diff-e5f6e2daef81ce1c3c4e9f7d992bd6ff9946b3b4d98a601e4d9573e5ef0cb07dR77


It’s fascinating how such a small, easily overlooked mistake can result in a severe attack surface.
At the same time, kudos to the developers who strive to build robust software. 👍

**Comment by ahmedazizkhelifi** [comment]
8 years of MongoBleed starting right here lol

**Comment by ctkqiang** [comment]
> 8 years of MongoBleed starting right here lol

superman-tell-me

**Pull request closed without merging**